Test deployment

mattjones315 opened the pull request:
This PR tracks changes necessary for deployment, including:

- Updated documentation (which can be checked [here](https://cassiopeia-lineage.readthedocs.io/en/testdeployment/index.html))
- Updated notebooks, including one for tree reconstruction
- Small bug fixes in ILPSolver and HybridSolver post-processing

Lioscro reviewed (changes requested):
Hey, Matt!
Thanks so much for taking the initiative to clean up everything for our next major release!
I have a few comments, and one more thing is (and I might have missed this somewhere), do we have any documentation for the command-line pipeline (like a description of the config file and how to run the pipeline from the comand-line?)

mattjones315 commented on cassiopeia/solver/HybridSolver.py:
```diff
@@ -326,7 +331,7 @@ def apply_bottom_solver(
             priors=cassiopeia_tree.priors,
         )
-        self.bottom_solver.solve(subtree)
+        self.bottom_solver.solve(subtree, logfile=logfile)
```
This actually occurred to me as well! Right now it works only because we've tried using the `ILPSolver` as the bottom-solver. I'll add `logfile` as an argument for `solve` so this should work across the board.

mattjones315 commented on docs/api/solver.rst:
```diff
@@ -0,0 +1,39 @@
+   solver.dissimilarity_functions.hamming_distance
+   solver.dissimilarity_functions.hamming_similarity_without_missing
+   solver.dissimilarity_functions.weighted_hamming_distance
+   solver.dissimilarity_functions.weighted_hamming_similarity
```
Yup! I'll add it.

mattjones315 commented on docs/conf.py:
```diff
@@ -154,21 +153,24 @@
     github_user="YosefLab",  # Username
     github_repo="Cassiopeia",  # Repo name
-    github_version="master",  # Version
+    github_version="refactor",  # Version
```
thanks for noting this

mattjones315 commented on docs/conf.py:
```diff
@@ -239,3 +241,52 @@ def setup(app):
+    option_spec = {
+        'methods': directives.unchanged,
+        'attributes': directives.unchanged
+    }
+
+    required_arguments = 1
+
+    @staticmethod
+    def get_members(obj, typ, include_public=None):
+        if not include_public:
+            include_public = []
+        items = []
+        for name in dir(obj):
+            try:
+                documenter = get_documenter(safe_getattr(obj, name), obj)
+            except AttributeError:
+                continue
+            if documenter.objtype == typ:
+                items.append(name)
+        public = [x for x in items if x in include_public or not x.startswith('_')]
+        return public, items
+
+    def run(self):
+        clazz = str(self.arguments[0])
+        try:
+            (module_name, class_name) = clazz.rsplit('.', 1)
+            m = __import__(module_name, globals(), locals(), [class_name])
+            c = getattr(m, class_name)
+            if 'methods' in self.options:
+                _, methods = self.get_members(c, 'method', ['__init__'])
+
+                self.content = ["~%s.%s" % (clazz, method) for method in methods if not method.startswith('_')]
+            if 'attributes' in self.options:
+                _, attribs = self.get_members(c, 'attribute')
+                self.content = ["~%s.%s" % (clazz, attrib) for attrib in attribs if not attrib.startswith('_')]
+        finally:
+            return super(AutoAutoSummary, self).run()
+
+def setup(app):
+    app.add_directive('autoautosummary', AutoAutoSummary)
```
Yeah I believe this is from Pandas

mattjones315 commented on notebooks/preprocess.ipynb:
```diff
@@ -54,23 +54,9 @@
   },
```
Thank you!

I've been meaning to update this notebook. Having some plots and common gotchas would also be helpful.

mattjones315 commented on .gitignore:
```diff
@@ -3,10 +3,13 @@
 *.pyc
 *.so
 _build
+<<<<<<< HEAD
+=======
```
thanks for catching this

mattjones315 commented on MANIFEST.in:
```diff
@@ -1,5 +1,4 @@
-include cassiopeia/TreeSolver/simulation_tools/*.pyx
+include cassiopeia/preprocess/*.pyx
+include cassiopeia/solver/*.pyx
+include *.toml
```
Looks like this doesn't break anything

mattjones315 commented on cassiopeia/solver/ILPSolver.py:
```diff
@@ -8,8 +8,12 @@
+try:
+    import gurobipy
+except ModuleNotFoundError:
+    warnings.warn("Gurobi not found. You must install Gurobi & gurobipy from source.")
```
That's a good question - I'm going to put this somewhere else in the file and throw an error.

mattjones315 commented:
Thanks for the feedback @Lioscro! I think I've addressed all of your comments, except about the documentation about the config file and command-line control of the preprocessing pipeline. What do you think of adding this to the `preprocess.ipynb`?

richardyz98 commented on cassiopeia/solver/HybridSolver.py:
```diff
@@ -310,7 +315,7 @@ def apply_bottom_solver(
             character_matrix = cassiopeia_tree.character_matrix.copy()
 
-        subproblem_character_matrix = character_matrix.loc[samples]
+        subproblem_character_matrix = character_matrix.loc[samples].drop_duplicates()
```
Wouldn't a bottom solver that already handles duplicates be able to handle this anyways?

mattjones315 replied:
Sure, let's do that.

richardyz98 reviewed (changes requested):
Hi Matt,
Things look great! 

I wonder if regarding the HybridSolver, we should leave duplicate processing to the bottom solver. What do you think?

mattjones315 commented on cassiopeia/solver/HybridSolver.py:
```diff
@@ -178,7 +181,10 @@ def solve(
-        samples_tree = self.__append_sample_names(tree, character_matrix)
+        samples_tree = self.__append_sample_names(tree, character_matrix, node_name_generator)
+
+        leaves = [n for n in samples_tree if samples_tree.out_degree(n) == 0]
```
Nope, good catch!

Lioscro reviewed (approved):
LGTM

richardyz98 reviewed (approved):
Great!

mattjones315 commented:
Thanks guys! Merging it in now.
